numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1192,150 @@ def infer_shapes(
+        self.alpha, self.tau = promote_shapes(alpha, tau, shape=batch_shape)
+        self.is_sparse = is_sparse
+
+        super(CAR, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+    def sample(self, key, sample_shape=()):
+        if self.is_sparse:
+            if not sparse.issparse(self.W):
+                assert isinstance(self.W, np.ndarray)
+                W = self.W
+            else:
+                W = sparse.csr_matrix(self.W)
+                W = np.asarray(W.todense())
+        else:
+            assert not sparse.issparse(self.W)
+            W = self.W
+
+        D = W.sum(axis=-1, keepdims=True) * jnp.eye(W.shape[-1])
+        tau = jnp.expand_dims(self.tau, (-2, -1))
+        alpha = jnp.expand_dims(self.alpha, (-2, -1))
+        P = tau * (D - alpha * W)
+        mvn = MultivariateNormal(self.loc, precision_matrix=P)
+        return mvn.sample(key, sample_shape=sample_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        phi = value - self.loc
+
+        if self.is_sparse:
+            if not sparse.issparse(self.W):
+                assert isinstance(self.W, np.ndarray)
+            W = sparse.csr_matrix(self.W)
+
+            D = np.asarray(W.sum(axis=-1)).squeeze()
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W.multiply(D_rsqrt @ D_rsqrt[:, np.newaxis]).toarray()
+
+            lam = np.linalg.eigvalsh(W_scaled)
+
+            W = BCOO.fromdense(W.todense())
+
+        else:
+            assert not sparse.issparse(self.W)
+            W = self.W
+
+            D = W.sum(axis=-1)
+            D_rsqrt = D ** (-0.5)
+
+            W_scaled = W * (D_rsqrt * D_rsqrt[:, jnp.newaxis])
+
+            lam = jnp.linalg.eigvalsh(W_scaled)
+
+        n = D.shape[-1]
+
+        logtau = n * jnp.log(self.tau)
+        logdet = jnp.log1p(-jnp.expand_dims(self.alpha, -1) * lam).sum(-1)
+
+        logquad = self.tau * jnp.sum(
+            phi * (D * phi - jnp.expand_dims(self.alpha, -1) * W @ phi), -1
+        )
+
+        return -0.5 * (logtau + logdet + logquad)
+
+    @property
+    def mean(self):
+        return self.loc
+
+    @property
+    def precision_matrix(self):
+        if self.is_sparse:
+            if not sparse.issparse(self.W):
+                assert isinstance(self.W, np.ndarray)
+                W = self.W
+            else:
+                W = sparse.csr_matrix(self.W)
+                W = np.asarray(W.todense())
+        else:
+            assert not sparse.issparse(self.W)
+            W = self.W
+
+        D = W.sum(axis=-1, keepdims=True) * jnp.eye(W.shape[-1])
+        tau = jnp.expand_dims(self.tau, (-2, -1))
+        alpha = jnp.expand_dims(self.alpha, (-2, -1))
+        return tau * (D - alpha * W)
+
+    @staticmethod
+    def infer_shapes(loc, alpha, tau, W):
+        event_shape = jnp.shape(W)[-1:]
```

Comment:
This already assumes the inputs are shapes, so you can do `event_shape = W[-1:]`